numpy/f2py/cfuncs.py

```diff
@@ -1007,36 +1007,39 @@
     PyObject* tmp = NULL;
     if (PyFloat_Check(obj)) {
         *v = PyFloat_AsDouble(obj);
-        return !(*v == -1.0 && PyErr_Occurred());
+        if (*v == -1.0) { goto error; }
+        return 1;
     }
 
     tmp = PyNumber_Float(obj);
     if (tmp) {
         *v = PyFloat_AsDouble(tmp);
         Py_DECREF(tmp);
-        return !(*v == -1.0 && PyErr_Occurred());
+        if (*v == -1.0){ goto error; }
+        return 1;
```

Comment:
This felt cleaner to me; whenever an error occurs `PyFloat_AsDouble` returns `-1` and the error should really just be handled at the end.

Reply:
I would expand it to (but maybe f2py does it similar):
```
if (*v == -1 && PyErr_Occurred()) {
    goto error;
}
```
In any case, `-1` is still a valid value if no error is set, so you need to check it before the `goto` (or return).  NumPy proper uses an `error_occurred(*v)` macro for this also.

Sometimes using goto generally is easier to read, here it seems also fine to me to just write:
```
if (*v == -1 && PyErr_Occurred()) {
    return 0;
}
return 1;
```
since there is no (relevant) cleanup here.

Reply:
> In any case, -1 is still a valid value if no error is set, so you need to check it before the goto (or return).

Yikes, yup, that makes sense, fixed.

>  NumPy proper uses an error_occurred(*v) macro for this also.
I wasn't able to find one which takes an argument but I might have been looking in the wrong place..

> since there is no (relevant) cleanup here.

Isn't the block at the end used to set a more informative error string?

Reply:
> I wasn't able to find one which takes an argument but I might have been looking in the wrong place..

Ah, the name was `error_converting(x)`, but I suspect it is not part of the public headers.

> Isn't the block at the end used to set a more informative error string?

Ohh, the code replaces the text/string, but not the error type.

Honestly, it is not really a pattern I am used to seeing.  Most of the errors seem fairly "severe" in any case: if we already checked for float, getting the value should not fail.

Writing it new, I would consider chaining errors if I think it is particularly important to give a more informative message.  Here my gut feeling would be to not bother replacing the text, but the other branches do anyway, so either seems fine.